agent/lib_php_amqplib.c

```diff
@@ -0,0 +1,770 @@
+    /* re-use start time from auto_segment started in func_begin */
+    message_segment->start_time = auto_segment->start_time;
+    nr_segment_message_end(&message_segment, &message_params);
+  }
+
+end:
+  /*
+   * Because we had to strdup values to persist them beyond
+   * NR_PHP_WRAPPER_CALL, now we destroy them. There isn't a separate function
+   * to destroy all since some of the params are string literals and we don't
+   * want to strdup everything if we don't have to. RabbitMQ basic_publish
+   * PHP 7.x will only strdup server_address, destination_name, and
+   * messaging_destination_routing_key.
+   */
+  UNDO_PERSISTENCE(message_params.server_address);
+  UNDO_PERSISTENCE(message_params.destination_name);
+  UNDO_PERSISTENCE(message_params.messaging_destination_routing_key);
+}
+NR_PHP_WRAPPER_END
+
+/*
+ * Purpose : A wrapper to instrument the php-amqplib basic_get.  This
+ * retrieves values to populate a message segment.  If
+ * newrelic.distributed_tracing_exclude_newrelic_header is false, it will also
+ * retrieve the DT headers and, if applicable, apply to the txn.
+ *
+ * PhpAmqpLib\Channel\AMQPChannel::basic_get
+ * Direct access to a queue if no message was available in the queue, return
+ * null
+ *
+ * @param string $queue
+ * @param bool $no_ack
+ * @param int|null $ticket
+ * @throws \PhpAmqpLib\Exception\AMQPTimeoutException if the specified
+ * operation timeout was exceeded
+ * @return AMQPMessage|null
+ */
+NR_PHP_WRAPPER(nr_rabbitmq_basic_get) {
+  zval* amqp_queue = NULL;
+  zval* amqp_exchange = NULL;
+  zval* amqp_routing_key = NULL;
+  zval* amqp_connection = NULL;
+  nr_segment_t* message_segment = NULL;
+  zval** retval_ptr = NR_GET_RETURN_VALUE_PTR;
+
+  nr_segment_message_params_t message_params = {
+      .library = RABBITMQ_LIBRARY_NAME,
+      .destination_type = NR_MESSAGE_DESTINATION_TYPE_EXCHANGE,
+      .message_action = NR_SPANKIND_CONSUMER,
+      .messaging_system = RABBITMQ_MESSAGING_SYSTEM,
+  };
+
+  (void)wraprec;
+
+  amqp_queue = nr_php_get_user_func_arg(1, NR_EXECUTE_ORIG_ARGS);
+  if (nr_php_is_zval_non_empty_string(amqp_queue)) {
+    /* For consumer, this is queue name. */
+    message_params.destination_name
+        = ENSURE_PERSISTENCE(Z_STRVAL_P(amqp_queue));
+  }
+
+  amqp_connection = nr_php_get_zval_object_property(
+      nr_php_execute_scope(execute_data), "connection");
+  /*
+   * In PHP 7.x, the following will create a strdup in
+   * message_params.server_address that needs to be freed.
+   */
+  nr_php_amqplib_get_host_and_port(amqp_connection, &message_params);
+
+  /* Compatibility with PHP 7.x */
+  NR_PHP_WRAPPER_CALL;
+
+  if (NULL == auto_segment) {
+    /*
+     * Must be checked after PHP_WRAPPER_CALL to ensure txn didn't end during
+     * the call.
+     */
+    goto end;
+  }
+  /*
+   *The retval should be an AMQPMessage. nr_php_is_zval_* ops do NULL checks
+   * as well.
+   */
+  if (nr_php_is_zval_valid_object(*retval_ptr)) {
```

Comment:
Check if `retval_ptr` is not `NULL` before dereferencing it:
```suggestion
  if (NULL != retval_ptr && nr_php_is_zval_valid_object(*retval_ptr)) {
```

Reply:
Since `nr_rabbitmq_basic_get` is used as both `special_instrumentation` and `special_instrumentation_clean` callbacks, `retval_ptr` will be NULL in the latter case (which is when `PhpAmqpLib\Channel\AMQPChannel::basic_get` throws an exception).

Reply:
`NULL` checking in `nr_php_is_zval_valid_object` is done after `retval_ptr` is dereferenced (`*retval_ptr`) which will yield undefined behavior (`SIGSEGV`) if `retval_ptr`'s value is `NULL`.